### DISCUSSION TITLE
Is asyncssh the solution of my problem?

### DISCUSSION THREAD

**Discussion by davidoiknine** [ronf/asyncssh#475] [Q&A] [answered]
Hello everyone,

I created a script where I send 15 000 commands to 70 switchs, so around 200 commands per switch to fetch data from them. 
First, I create one SSH connection with a jump server. Then, for each switch, I create one SSH connection via jumpSSH (via the jump server) and once the connection is established, I'm sending 200 commands with remote_session.get_cmd_output using multithreading.

```
def main_loop():
    gateway_session = SSHSession('gateway.example.com',
                                 'my_user', password='my_password').open()
    for switch in result.keys():
        remote_session = gateway_session.get_remote_session(switch["ip"],
                                                    password='my_password2')
        with ThreadPoolExecutor(max_workers=8) as executor:
            for command in result[switch]["commands"]:
                executor.submit(launch_command, command, remote_session)

def launch_command(command, remote_session):
    process = remote_session.get_cmd_output(command)                
```

My script is taking 30 minutes to run and to fetch all the data I want from the switches.

Do you think asyncssh is a good tool for what I'm doing and will the performance of asyncssh be better than what I'm using today?

Thank you,

**Comment by ronf** [comment]
The exact performance will depend on how quickly the switches respond to your commands, but I did a quick test here of opening 75 connections to "localhost", with each connection sequentially starting 200 sessions running "ls" and collecting up all the results. To avoid running into simultaneous authentication limits on localhost, I added a semaphore that limited the code to starting at most 10 connections at once, but you probably wouldn't need that if you were connecting to different target systems.

Here's some example code:

```python
import asyncio, asyncssh

sem = asyncio.Semaphore(10)
targets = 75*[{'host': 'localhost', 'commands': 200*['ls']}]

async def run_client(host, commands):
    async with sem:
        conn = await asyncssh.connect(host)

    async with conn:
        return [await conn.run(command) for command in commands]

async def run_multiple_clients():
    tasks = (run_client(target['host'], target['commands']) for target in targets)
    results = await asyncio.gather(*tasks, return_exceptions=True)

asyncio.run(run_multiple_clients())
```

Even when running all of this in a single event loop (and thus only being able to take advantage of a single core), my test finished in just a little over 1 minute.

Here's a variation that uses multiple threads, with an event loop per thread. Interestingly, it's actually a bit slower than the single event loop version, even though it did take advantage of more than one CPU core.

```python
import asyncio, asyncssh
from concurrent.futures import ThreadPoolExecutor

targets = 75*[{'host': 'localhost', 'commands': 200*['echo foo']}]

async def run_client(host, commands):
    async with asyncssh.connect(host) as conn:
        return [await conn.run(command) for command in commands]

def run_async(target):
    return asyncio.run(run_client(target['host'], target['commands']))

if __name__ == '__main__':
    with ThreadPoolExecutor(10) as executor:
        results = executor.map(run_async, targets)
```

To get real benefit here, it might be necessary to use multiple processes instead, so the Python global interpreter lock doesn't get in the way of concurrency. However, attempting to switch to ProcessPoolExecutor and processing results locally in each process to avoid having to return them to the main process only shaved off 10 seconds or so, and it used a lot more total CPU. By far, the first version above was the most efficient.

It's hard to know with your actual target hosts whether AsyncSSH will be faster than your current solution or not, but it's certainly capable of handling this kind of workload without a problem. The code is also pretty compact.

**Comment by davidoiknine** [comment]
Hello Ron,

Thank you so much for your prompt reply.
I am currently testing the module.

I had a bit of trouble implementing a jump server connection at first, but I finally figured out how to do it.
So I'm modifying my script to make it work with yours.

I will let you know as soon as I finish it.

Thank you
David

**Reply by davidoiknine** [reply]
Hello Ron,

I managed to make it work for the first step of my script. I'm iterating over 70 switches and I send only one command per switch. All of this, connecting via a jump server. It took 9 seconds to proceed.

Now, I'm trying to make the second step.
In the step 1, I fetch mac, ip addresses and vlans of all hosts that are attached locally to each switch.
And I've built the dictionnary **leafs** like that:

```
{
    "hostname1": {
        "leaf_ip": "10.234.12.4",
        "hosts": [
            {
                "ip": "10.42.78.59",
                "mac": "ab:cd:ef:gh:ij:01",
                "vlan": "107"
            },
            {
                "ip": "10.42.78.60",
                "mac": "ab:cd:ef:gh:ij:02",
                "vlan": "161"
            },
            {
                "ip": "10.42.78.61",
                "mac": "ab:cd:ef:gh:ij:03",
                "vlan": "161"
            }
        ]
    },
    "hostname2": {
        "leaf_ip": "10.234.12.5",
        "hosts": [
            {
                "ip": "10.42.78.62",
                "mac": "ab:cd:ef:gh:ij:04",
                "vlan": "107"
            },
            {
                "ip": "10.42.78.64",
                "mac": "ab:cd:ef:gh:ij:06",
                "vlan": "161"
            }
        ]
    },
    "hostname3": {
        "leaf_ip": "10.234.12.6",
        "hosts": [
            {
                "ip": "10.42.78.65",
                "mac": "ab:cd:ef:gh:ij:07",
                "vlan": "107"
            },
            {
                "ip": "10.42.78.67",
                "mac": "ab:cd:ef:gh:ij:09",
                "vlan": "161"
            }
        ]
    },
    "hostname4": {
        "leaf_ip": "10.234.12.6",
        "hosts": []
    }
}

```

I want now to iterate over each host to launch another command on the switch it is attached, this to get another information: The port on which the host is attached to the switch.

In order to do this, here what I am doing:

```
sem = asyncio.Semaphore(10)
async def get_port_one(mac, conn):
    async with conn:
        return await conn.run('sh mac address-table address ' + mac + ' | json')

async def get_port_all(leafs):
    async with sem:
        tunnel = await asyncssh.connect('xx.xx.xx.xx', username='admin', password='xxxxxxxxx', known_hosts=None)
    tasks = []
    for hostname in leafs.keys():
        leaf_ip = leafs[hostname]["leaf_ip"]
        async with sem:
            conn = await tunnel.connect_ssh(leaf_ip, username='admin', password='xxxxxxxx', known_hosts=None)
        for host in leafs[hostname]["hosts"]:
            mac = host["mac"]
            task = get_port_one(mac, conn)
            tasks.append(task)
    results = await asyncio.gather(*tasks, return_exceptions=True)
    for result in results:
        print(result)

def main():
    asyncio.run(get_port_all(leafs))
    print(f'Temps total: {time() - start}')


if __name__ == '__main__':
    main()

```


The problem is that I get result for only 97 hosts over 5000. For the others, I get two kind of errors:

- Session request failed
OR
- open failed
OR
- SSH connection closed

I think I did a mistake in the way I wrote my script.
At the beginning, I didn't used any semaphore (as in my first step), but for the second step, I added it but I'm not sure it is well implemented. Also, how could I know how much connections I can make at the same time please?
Can you please help me with this?

Thank you so much
David

**Reply by ronf** [reply]
Regarding jump hosts, doing a connect() to the jump host and then doing connect_ssh() on that tunnel is one open, but in that case you'll probably want to move the "async with sem" to only be used on the inner connect_ssh() connections, as you'll only have one connection to the jump host that all of the other connections are sharing. Also, I don't know if the Python asyncio semaphore is designed to be recursive -- acquiring it on the inner calls while you're already holding it at the outer level could be a problem.

There's a more fundamental issue here, though, which is that you are waiting for the calls to connect_ssh() to complete in your for loop over the leaf_ips, and you probably don't want to do that. You want to allow those connections to go in parallel, but with a limit enforced by the semaphore to avoid too many connections being opened at once. Note how in my example the "async with sem" is only around the connect() call, and the semaphore is released before the "async with conn" which uses the connection to send commands.

Once they are opened, you can then open sessions for your commands. You may not want to do those in parallel, though -- that'll depend whether the switches you are contacting can handle parallel commands or not. In my example, I did the initial connects in parallel (while holding the semaphore), but then sent the commands to each target serially, to avoid overloading the target with too many sessions at once. You could possible do this in parallel with another set of semaphore, but I'd suggest starting with sending the commands serially and getting that to work before you attempt creating a task per command like you're doing now. Instead, just have a single task which does the connect_ssh() to a leaf and then does a for loop to serially send commands to that switch.

In your current code, it looks like you are creating a task per command with no limit on the number of those you open in parallel, so that may explain some of the errors you are seeing. Between that and reusing the semaphore, I'm not really sure that's going to do any limiting at all.

Regarding getting the command results, the SSHCompletedProcess returned by run() has a "stdout" member with the command output. It also has an exit_status, but my guess is that switches probably won't set that. So, you'll have to probably need to parse the output to determine the success of the command and extract any result you want from it.

As for the number of connections you can make in parallel, that depends on the server you are connecting to. There is usually one limit on number of simultaneous authentications allowed (MaxStartups in OpenSSH), and there may be some limit on the total number of connections, but that's usually much higher. There is also generally a limit on the number of sessions any single connection can support (MaxSessions in OpenSSH), so you may run into that limit on the jump host. If you hit that, you may need to make multiple independent connections to the jump host to stay under that limit. By default, you need to stay under 10 connections in authentication at once to avoid OpenSSH returning failures, and stay under 10 simultaneous sessions. The switches you are talking to may limit this even more, so you'll simply have to experiment (or check the switch documentation).

**Reply by davidoiknine** [reply]
Hello Ron,

Thank you for your reply.

Yes you completely right, actually I'm working switch per switch whereas I should send commands at the same time to several switches in parallel.

The issue I have is that each command is different because each command has the mac of the host like that:
'sh mac address-table address ' + mac
So I don't know how can I make my code in order to send different commands at the same time to different switches.

You said:
"Instead, just have a single task which does the connect_ssh() to a leaf and then does a for loop to serially send commands to that switch."

You mean I should create one task for each switch? Then I do a for loop to add inside the different commands?
I'm sorry I didn't get well this part.

Do you mean something like that?

```
sem = asyncio.Semaphore(10)
async def get_port_one(conn, leafs, hostname):
    for host in leafs[hostname]["hosts"]:
        async with conn:
            mac = host["mac"]
            await conn.run('sh mac address-table address ' + mac + ' | json')
            
async def get_port_all(leafs):
    tunnel = await asyncssh.connect('xx.xx.xx.xx', username='admin', password='xxxxxxxxxx', known_hosts=None)
    tasks = []
    for hostname in leafs.keys():
        leaf_ip = leafs[hostname]["leaf_ip"]
        async with sem:
            conn = await tunnel.connect_ssh(leaf_ip, username='admin', password='xxxxxxxxxx', known_hosts=None)
        task = get_port_one(conn, leafs, hostname)
        tasks.append(task)
    results = await asyncio.gather(*tasks, return_exceptions=True)
    for result in results:
        print(result)
```

Thank you,

**Comment by ronf** [comment]
That's getting closer, but you'll want to move the connect_ssh() call and the "async with sem" around it to be the first lines in get_port_one(). That way, that code runs as part of the tasks you are creating and running in parallel, instead of the connects running serially in the for loop in get_port_all() as the tasks are being created.

Once you've done that, I think the main remaining issue will be that you may hit the maximum number of allowed sessions on the jump host. To fix that, you'd need to call asyncssh.connect() multiple times, limiting the number of calls to connect_ssh() made on any given tunnel. You can still reuse a tunnel object for multiple connect_ssh() calls, but just not ALL of the calls if you have a lot of switches. Moving the asyncssh.connect() inside the for loop and keeping a counter for the number of sessions you've opened on it would probably be the easiest thing. When the count hits 10, do a new call to asyncssh.connect(), replacing the previous tunnel object. The previous sessions you created should be enough to keep the previous tunnel connection from being garbage-collected even though you don't have an explicit reference to it. You'd be relying on garbage collection to clean it up later on when you are done, but since your program exits at that point, that's not a big concern.

You could also keep an explicit list of all the tunnel objects you open in this way, and call close() and await wait_closed() on them at the very end, if you wanted to cleanly free them.

**Reply by davidoiknine** [reply]
Hello Ron and thank you for your time.

So I've done all what you said and it all makes sense but I still receive errors in results:

```
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
None
None
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
None
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
SSH connection closed
None
None
None
None
None
None
None
None
None
None
```

Here is my code:

```
sem = asyncio.Semaphore(10)
async def get_port_one(tunnels, tunnel, leafs, hostname, leaf_ip, i):
    async with sem:
        conn = await tunnel.connect_ssh(leaf_ip, username='admin', password='xxxxxxxx', known_hosts=None)
    for host in leafs[hostname]["hosts"]:
        if i > 10:
            i = 0
            tunnel = await asyncssh.connect('xx.xx.xx.xx', username='admin', password='xxxxxxxx', known_hosts=None)
            tunnels.append(tunnel)
        async with conn:
            ip = host["ip"]
            i += 1
            await conn.run('sh bgp l2vpn evpn ' + ip + ' | json')
            
async def get_port_all(leafs):
    tunnels = []
    tasks = []
    i = 0
    tunnel = await asyncssh.connect('xx.xx.xx.xx', username='admin', password='xxxxxxxx', known_hosts=None)
    tunnels.append(tunnel)
    for hostname in leafs.keys():
        if i > 10:
            i = 0
            tunnel = await asyncssh.connect('xx.xx.xx.xx', username='admin', password='xxxxxxxx', known_hosts=None)
            tunnels.append(tunnel)
        print(hostname)
        leaf_ip = leafs[hostname]["leaf_ip"]
        task = get_port_one(tunnels, tunnel, leafs, hostname, leaf_ip, i)
        tasks.append(task)
    results = await asyncio.gather(*tasks, return_exceptions=True)
    for tunnel in tunnels:
        tunnel.close()
        await tunnel.wait_closed() 
    for result in results:
        print(result)
```

Maybe is it the fact that each task I'm appending to the tasks list is not just one task but multiple conn.run tasks?

Thank you very much,
David

**Reply by ronf** [reply]
This definitely looks better. Having a single task for each switch which runs the commands sequentially is a good first step. You could try and parallelize it further if the switch can handle multiple parallel sessions on a single connection, but this I'd try to get this version to work before you attempt parallelizing the conn.run() calls.

The semaphore looks good here now, but you may be running into limits on the ability for the jump host to handle more than 10 sessions on a single connection. Right now, the semaphore just protects against MaxStartups on a given switch but not MaxSessions on the jump host. To deal with that, you may need to open multiple jump host connections, to limit each connection to at most 10 sessions. Either that, or most the "for" loop in get_port_one() inside the "async with sem" block so that you only allow at most 10 sessions total to be running at once, delaying the remaining sessions until some earlier session completes running all of its commands and closes the jump host session.

**Reply by davidoiknine** [reply]
Hello Ron,

I am sorry for the delay in my answer. I was in vacation for a few weeks.
Good news, I finally managed to make it work thanks to you and the result is much better than multithreading (less than half the time).

I really would like to thank you for your time and your patience with me. 
Thanks a lot.

David

**Comment by ronf** [comment, marked as answer]
That's terrific, David -- glad to hear you got it working, and got such an improvement!

I'm happy to help - thanks for letting me know how you made out...

**Comment by davidoiknine** [comment]
Hello @ronf, I hope you are doing well.

I'm trying to use again asyncSSH to send several commands on the same equipment this time, but i'm having some trouble to do it.

I want to send the command ```dig -x  xx.xx.xx.xx +short```, with xx.xx.xx.xx an ip address.
I have 15000 ip addresses so I want to send the command 15000 times.

```
async def hostname(leafs):
    tasks = []
    async with asyncssh.connect(os.getenv("EQUIPMENT_IP"),
                        username='admin', password=os.getenv("EQUIPMENT_PASSWORD"), known_hosts=None) as conn:
        for hostname in leafs.keys():
            for host in leafs[hostname]["hosts"]:
                ip = host["ip"]
                task = await conn.run("dig -x " + ip + " +short", timeout=5)
                tasks.append(task)
    results = await asyncio.gather(*tasks, return_exceptions=True)        
    print(results)

asyncio.run(hostname(leafs))
```

When I'm running this script, it seems like an infinite loop and nothing happens.
Maybe do you see the problem in my script?
Also, am I forced to set a semaphore in this case?

Thank you,
David

**Reply by ronf** [reply]
As currently written, the code will need to finish running all 15,000 SSH sessions before you see any output. So, it might be working, but it may just take a very long time to complete. If you want to see the output as you go, you'd need to replace the `gather()` call with something like a for loop over `asyncio.as_completed()` of your tasks. That will return one task result at a time as they complete, allowing you to see results as they become available. This would look something like:

```python
    for task in asyncio.as_completed(tasks):
        result = await task
        print(result)
```

Note that the order of the results is unpredictable, though. If you are trying to associate results with specific requests, you may need to provide a little wrapper around `conn.run()` which can return the hostname/IP of each request along with the result returned.

**Reply by davidoiknine** [reply]
Thank you @ronf 
I tried to write a more simple and reproducible script.

```
import asyncio, asyncssh

ips = ["192.168.1.1", "192.168.1.2"]

async def get_hostname(ips):
    tasks = []
    conn = await asyncssh.connect(host="xx.xx.xx.xx", username='admin', password="xxxxxxxxxx", known_hosts=None)
    for ip in ips:
        task = await conn.run("dig -x " + ip + " +short", timeout=5)
        tasks.append(task)
    for task in asyncio.as_completed(tasks):
        result = await task
        print(result)


asyncio.run(get_hostname(ips))
```

And now, I get the following traceback:

```
Traceback (most recent call last):
  File "/home/david/vm-tracer/hostname.py", line 16, in <module>
    asyncio.run(hostname(ips))
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/home/david/vm-tracer/hostname.py", line 11, in hostname
    for task in asyncio.as_completed(tasks):
  File "/usr/lib/python3.10/asyncio/tasks.py", line 549, in as_completed
    todo = {ensure_future(f, loop=loop) for f in set(fs)}
  File "/usr/lib/python3.10/asyncio/tasks.py", line 549, in <setcomp>
    todo = {ensure_future(f, loop=loop) for f in set(fs)}
  File "/usr/lib/python3.10/asyncio/tasks.py", line 615, in ensure_future
    return _ensure_future(coro_or_future, loop=loop)
  File "/usr/lib/python3.10/asyncio/tasks.py", line 629, in _ensure_future
    raise TypeError('An asyncio.Future, a coroutine or an awaitable '
TypeError: An asyncio.Future, a coroutine or an awaitable is required
```

I don't know where this error comes from.
Thank you Ron

**Reply by ronf** [reply]
Ah - I see the issue now. There shouldn't be an `await` in front of the `conn.run()` call. You want to create the task but not actually block waiting for it inside that for-loop. Instead, you want to wait for all the tasks to be created before you start any of them, so they can all run in parallel.

Regarding your question about the semaphore, you will probably need to add that in as well, to avoid hitting the limit on the number of sessions you have open at once on a single SSH connection.

**Reply by davidoiknine** [reply]
Thank you @ronf. It's working now.
Fortunately, I don't need to set a semaphore, it's working without it.

But you're right, it is very long to proceed. Do you know why? Is there a way to accelerate the process?

Thank you,

**Reply by ronf** [reply]
Adding the semaphore may speed things up, as asking the target system to handle 15,000 simultaneous sessions may be contributing to the slowness. Running fewer of those commands in parallel may allow them to complete more quickly overall, even if that means some of the commands will be a bit delayed in starting up. If there are algorithms in the server that do things like a linear search to map channel numbers to channels, running with the semaphore should take less time to do the search.

Do you really need to be running this IP reverse lookup on the target devices? Generally speaking, that's just going to trigger the target to send queries to its DNS servers. If you know what those DNS servers are, perhaps you could do the reverse lookup from your client machine without getting the target machine involved. You wouldn't even have to use "dig" here -- you could use something like Python's `socket.getnameinfo()` if your client's DNS servers are the same as the target machine.

Also, one of the things which may be contributing to the slowness is that some DNS servers don't respond to DNS reverse lookup requests, causing the lookup to eventually time out. Are you seeing any examples of that, where the "dig" command returns no output for some of the lookups? Unfortunately, unless you own the DNS servers for all of the IPs you are looking up, this may not totally be under your control.

**Reply by davidoiknine** [reply]
Hello @ronf,

Thank you for your reply.
I'm now doing the treatment directly on my machine. I also use semaphore.
And I have two issues:

1) It's still very long
2) At the end of the treatment, after getting the results from all of the ip addresses, I get the same error as previously. 

```
Traceback (most recent call last):
  File "/home/david/vm-tracer/hostname.py", line 30, in <module>
    asyncio.run(hostname(ips))
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 641, in run_until_complete
    return future.result()
  File "/home/david/vm-tracer/hostname.py", line 24, in hostname
    for task in asyncio.as_completed(tasks):
  File "/usr/lib/python3.10/asyncio/tasks.py", line 549, in as_completed
    todo = {ensure_future(f, loop=loop) for f in set(fs)}
  File "/usr/lib/python3.10/asyncio/tasks.py", line 549, in <setcomp>
    todo = {ensure_future(f, loop=loop) for f in set(fs)}
  File "/usr/lib/python3.10/asyncio/tasks.py", line 615, in ensure_future
    return _ensure_future(coro_or_future, loop=loop)
  File "/usr/lib/python3.10/asyncio/tasks.py", line 629, in _ensure_future
    raise TypeError('An asyncio.Future, a coroutine or an awaitable '
TypeError: An asyncio.Future, a coroutine or an awaitable is required
```

And it doesn't display the total time.

Here is my final script:

```
import asyncio, asyncssh
import os
from time import time

start = time()
ips = [] // we assume the ips list is full of IP addresses

sem = asyncio.Semaphore(20)
async def hostname(ips):
    tasks = []
    for ip in ips:
        task = os.system("dig -x " + ip + " +short")
        tasks.append(task)
    for task in asyncio.as_completed(tasks):
        async with sem:
            result = await task
            print(result)

asyncio.run(hostname(ips))
print(f'Temps total: {time() - start}')
```

Thank you,

**Reply by ronf** [reply]
I'm a bit surprised this would work as-is, given that os.system() is not an async function. I'd expect it to fail when you did the "await task", if you waited long enough for it to finish appending all the tasks. That first part is going to be very slow, though, as it'll be doing all the os.system() calls in a blocking manner, one at a time.

You can also make it faster by not launching a separate process for each lookup, using the asyncio version of the getnameinfo() function to do reverse lookups. Here's an example:

```python
import asyncio
import os
from time import time

start = time()
ips = []  # we assume the ips list is full of IP addresses

async def lookup_host(loop, ip):
    return (await loop.getnameinfo((ip, 0), 0))[0]

sem = asyncio.Semaphore(20)
async def hostname(ips):
    loop = asyncio.get_event_loop()

    tasks = []
    for ip in ips:
        tasks.append(lookup_host(loop, ip))
    for task in asyncio.as_completed(tasks):
        async with sem:
            result = await task
            print(result)

asyncio.run(hostname(ips))
print(f'Temps total: {time() - start}')
```

On my machine here, 15,000 lookups of names owned by a local resolver complete in a little under 2 seconds, but there might be some caching involved in that. So, with a larger number of unique IPs, it might take a bit longer. It should be much faster than using AsycnSSH or os.system(), though, even if you were to fix the os.system() calls to happen in parallel.

**Reply by davidoiknine** [reply]
Hi Ron,

It's very weird, the script is working but taking very long time at the beginning before displaying all the results in one second.
It took 190 seconds in total.

Thanks

**Reply by ronf** [reply]
In cases where no reverse lookup exists for a given IP address, this script will return back the original IP, unlike "dig" which ends up returning an empty string.

If the script stops generating output, it may be because it is waiting for a DNS response. It's common for some DNS servers to not respond to reverse lookups, leading to clients waiting a long time and then eventually timing out. That could be part of why you were seeing slowness in the first place.

Looking at the positioning of the semaphore, I also think that's in the wrong place now that as_completed() is being used. It should probably be more like:

```python
import asyncio
import os
from time import time

start = time()
ips = []  # we assume the ips list is full of IP addresses
sem = asyncio.Semaphore(20)

async def lookup_host(loop, ip):
    async with sem:
        return (await loop.getnameinfo((ip, 0), 0))[0]

async def hostname(ips):
    loop = asyncio.get_event_loop()

    tasks = []
    for ip in ips:
        tasks.append(lookup_host(loop, ip))
    for task in asyncio.as_completed(tasks):
        result = await task
        print(result)

asyncio.run(hostname(ips))
print(f'Temps total: {time() - start}')
```

However, this won't necessarily help if some of the DNS servers you are going to take a long time to respond (or never respond at all). Are all of the addresses you are looking up "internal" addresses, with their reverse lookup owned by local DNS servers?
